rclcpp_components/include/rclcpp_components/component_manager_isolated.hpp

```diff
@@ -0,0 +1,61 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#ifndef RCLCPP_COMPONENTS__COMPONENT_MANAGER_ISOLATED_HPP__
+#define RCLCPP_COMPONENTS__COMPONENT_MANAGER_ISOLATED_HPP__
+
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+#include <unordered_map>
+
+#include "rclcpp_components/component_manager.hpp"
+
+
+namespace rclcpp_components
+{
+/// ComponentManagerIsolated uses dedicated single-threaded executors for each components.
+class ComponentManagerIsolated : public rclcpp_components::ComponentManager
+{
+  using rclcpp_components::ComponentManager::ComponentManager;
+
+public:
+  ~ComponentManagerIsolated();
+
+protected:
+  RCLCPP_COMPONENTS_PUBLIC
+  void
+  on_load_node(
+    const std::shared_ptr<rmw_request_id_t> request_header,
+    const std::shared_ptr<LoadNode::Request> request,
+    std::shared_ptr<LoadNode::Response> response) override;
+  RCLCPP_COMPONENTS_PUBLIC
+  void
+  on_unload_node(
+    const std::shared_ptr<rmw_request_id_t> request_header,
+    const std::shared_ptr<UnloadNode::Request> request,
+    std::shared_ptr<UnloadNode::Response> response) override;
+
+private:
+  std::unordered_map<uint64_t, std::shared_ptr<rclcpp::Executor>> dedicated_executors_;
+  std::unordered_map<uint64_t, std::thread> dedicated_threads_;
+  std::unordered_map<uint64_t, std::promise<void>> dedicated_promises_;
```

Comment:
Maybe you can create a custom struct for the three values (executor, thread, promise) and create only one unordered map.
You could also use a tuple instead of defining a custom struct (defining a custom struct makes the code more readable than using tuples).

Reply:
Agreed. A custom struct is the cleanest way